src/storage/gpu_swap_history.cc

```diff
@@ -287,7 +343,55 @@ void MemoryHistory::StopIteration() {
   }
 }
 
-} // namespace mxnet
-
+double MemoryHistory::LCS_Similarity(std::vector<MemRecord>& base,
+                                     std::vector<MemRecord>& target) {
+  std::vector<size_t> *prev_table, *curr_table, *temp;
+  size_t size1 = base.size();
+  size_t size2 = target.size();
+  prev_table = new std::vector<size_t>(size2 + 1, 0);
+  curr_table = new std::vector<size_t>(size2 + 1, 0);
+  for (size_t i = 1; i <= size1; i++) {
+    for (size_t j = 1; j <= size2; j++) {
+      if (base[i - 1].handle_id == target[j - 1].handle_id) {
+        (*curr_table)[j] = (*prev_table)[j - 1] + 1;
+      } else {
+        (*curr_table)[j] = std::max((*curr_table)[j - 1], (*prev_table)[j]);
+      }
```

Comment:
I'm pretty sure the original implementation is correct since it matches the recursive formula for LCS proved in Introduction to Algorithms. You should be able to prove that table[i-1][j-1] + 1 >= table[i - 1][j] and table [i][j - 1] for the case that str1[i] == str2[j] by using contraction. This is the wiki version of the formula. https://en.wikipedia.org/wiki/Longest_common_subsequence_problem

Reply:
I see.